pyiron_atomistics/vasp/base.py

```diff
@@ -1843,15 +1909,70 @@ def __init__(self):
         self.incar = Incar(table_name="incar")
         self.kpoints = Kpoints(table_name="kpoints")
         self.potcar = Potcar(table_name="potcar")
-
+        self.options = VaspSpecificOptions()
         self._eddrmm = "warn"
 
+        self.structure = None
+        self.__idx_user_to_pyiron = []
+        self.__idx_pyiron_to_user = []
+
+    @property
+    def _idx_user_to_pyiron(self):
+        if self.structure is None:
+            return None
+        if len(self.__idx_user_to_pyiron) == 0:
+            self._map_pyiron_to_user_idx()
+        return self.__idx_user_to_pyiron
+
+    @property
+    def _idx_pyiron_to_user(self):
+        if self.structure is None:
+            return None
+        if len(self.__idx_pyiron_to_user) == 0:
+            self._map_pyiron_to_user_idx()
+        return self.__idx_pyiron_to_user
+
+    def _map_pyiron_to_user_idx(self):
+        """
+        This writes the indices maps for user->pyiron (_idx_user_to_pyiron), and pyiron->user (_idx_pyiron_to_user)
+        This looks in Input.options for the allow_structure_reordering boolean value, and decides to return either
+        1. Sorting map when species-based reordering is allowed (default behaviour of pyiron)
+        or
+        2. Sorting map when sorting is forbidden, in which case naive map (1:1, 2:2 etc.) is returned.
+        """
+        if self.options.allow_structure_reordering:
+            atom_numbers = self.structure.get_number_species_atoms()
+
+            _idx_user_to_pyiron = list()
+            for species in atom_numbers.keys():
+                indices = self.structure.select_index(species)
+                for i in indices:
+                    _idx_user_to_pyiron.append(i)
+            self.__idx_user_to_pyiron = np.array(_idx_user_to_pyiron)
+
+            _idx_pyiron_to_user = np.array([0] * len(_idx_user_to_pyiron))
+            for i, p in enumerate(_idx_user_to_pyiron):
+                _idx_pyiron_to_user[p] = i
+            self.__idx_pyiron_to_user = _idx_pyiron_to_user
+        else:
+            self.__idx_user_to_pyiron = np.arange(len(self.structure))
+            self.__idx_pyiron_to_user = np.arange(len(self.structure))
+
     def write(self, structure, modified_elements, directory=None):
         """
         Writes all the input files to a specified directory
 
         Args:
-            structure (atomistics.structure.atoms.Atoms instance): Structure to be written
+            structure (atomistics.structure.atoms.Atoms instance): Structure (unsorted) to be written
+
+            The structure obj being fed into this fn at .write_input() function is sorted;
```

Comment:
> This documentation makes assumptions about the behaviour of another function. What's more, it makes assumptions about a function that calls this function, which I feel is very easy to lose track of. In contrast, if some of this same discussion were in write_input, at least there we'd be discussing assumptions about a function we're calling.

This is mostly just me trying to not completely rewrite the entire module and trying to preserve the original as much as possible xD

If it were up to me, I would have completely rewritten large parts of this module, but this is what I have inherited, and I thought some might prefer that I leave it the way it was. (Even while being somewhat self-aware of my hubris, I still think that I could do a much better job with a complete re-write for achieving what some parts of this module is trying to do.) 

I don't like it any more than you do. The logic is mostly written in the docstring for me to understand what exactly was going on in the (somewhat convoluted) code, and for any future unlucky people that are trying to work with this module.

> The bit about "if you call it manually and want to be consistent with write_input is good, but I see we use a private variable. Should we make _idx_X_to_Y public?

it was originally public in an earlier commit, but @pmrv suggested it private since he didn't want it polluting the tab complete. So I guess we should have a throwdown xD. His complaint was mostly the naming, but I think `idx_user_to_pyiron` and vice-versa make great variable names since it makes it pretty clear what's exactly going on. 

> I don't think we need abbreviations like "obj" and "fn" in the docs.

Sure, I'll remove them and put in full words